bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -499,176 +576,163 @@ void createNewLog() throws IOException {
-        mapInfo.putLong(ledgerMapOffset);
-        mapInfo.putInt(numberOfLedgers);
-        mapInfo.flip();
-        entryLogChannel.fileChannel.write(mapInfo, LEDGERS_MAP_OFFSET_POSITION);
-    }
 
     /**
      * An allocator pre-allocates entry log files.
      */
     class EntryLoggerAllocator {
 
         private long preallocatedLogId;
-        private Future<BufferedLogChannel> preallocation = null;
+        Future<BufferedLogChannel> preallocation = null;
         private ExecutorService allocatorExecutor;
-        private final Object createEntryLogLock = new Object();
-        private final Object createCompactionLogLock = new Object();
 
         EntryLoggerAllocator(long logId) {
             preallocatedLogId = logId;
             allocatorExecutor = Executors.newSingleThreadExecutor();
         }
 
-        BufferedLogChannel createNewLog() throws IOException {
-            synchronized (createEntryLogLock) {
-                BufferedLogChannel bc;
-                if (!entryLogPreAllocationEnabled){
-                    // create a new log directly
-                    bc = allocateNewLog();
-                    return bc;
-                } else {
-                    // allocate directly to response request
-                    if (null == preallocation){
-                        bc = allocateNewLog();
+        synchronized long getPreallocatedLogId(){
+            return preallocatedLogId;
+        }
+
+        synchronized BufferedLogChannel createNewLog() throws IOException {
+            BufferedLogChannel bc;
+            if (!entryLogPreAllocationEnabled || null == preallocation) {
+                // initialization time to create a new log
+                bc = allocateNewLog();
+            } else {
+                // has a preallocated entry log
+                try {
+                    /*
+                     * both createNewLog and allocateNewLog are synchronized on
+                     * EntryLoggerAllocator.this object. So it is possible that
+                     * a thread calling createNewLog would attain the lock on
+                     * this object and get to this point but preallocation
+                     * Future is starving for lock on EntryLoggerAllocator.this
+                     * to execute allocateNewLog. Here since we attained lock
+                     * for this it means preallocation future must have either
+                     * completed creating new log or still waiting for lock on
+                     * this object to execute allocateNewLog method. So we
+                     * should try to get result of the future without waiting.
+                     * If it fails with TimeoutException then call
+                     * allocateNewLog explicitly since we are holding the lock
+                     * on this anyhow.
+                     *
+                     */
+                    bc = preallocation.get(0, TimeUnit.MILLISECONDS);
```

Comment:
i don't think it is correct for createNewLog to be guarded by createEntryLogLock and createNewLogForCompaction to be guarded by createCompactionLogLock when both call non-synchronized allocateNewLog method (which does operations on shared state, for eg:  preallocatedLogId) and also for clarity sake it would be easier to have all these methods synchronized on same lock (this). 

(same comment as I mentioned in code)
now since both createNewLog and allocateNewLog are synchronized on EntryLoggerAllocator.this object. So it is possible that a thread calling createNewLog would attain the lock on this object and get to this point but preallocation Future is starving for lock on EntryLoggerAllocator.this to execute allocateNewLog. Here since we attained lock for this it means preallocation future must have either completed creating new log or still waiting for lock on this object to execute allocateNewLog method. So we should try to get result of the future without waiting. If it fails with TimeoutException then call allocateNewLog explicitly since we are holding the lock on this anyhow.

Reply:
@reddycharan okay, I remembered this context. however does this change need to be here for introducing EntryLogManager? If you can defer this change to a subsequent change, that would be better, also it is good for people understand the change here in future when looking back at the git history.

Reply:
> i don't think it is correct for createNewLog to be guarded by createEntryLogLock and createNewLogForCompaction to be guarded by createCompactionLogLock when both call non-synchronized allocateNewLog method

I think we discussed that. Technically it is not a correctness issue. because the way how transactional compaction works.

The main question I have here, `preallocation.get(0, TimeUnit.MILLISECONS)`. if the allocation is still happening, this call will timeout immediately, and you go directly allocate a new entry log file again, which leaves an entry log with a smaller log id used later. this is a potential bug to me, because single-entry-log implementation assumes log id increasing at some extension (for both checkpoint, gc, compaction). 

regarding your point "but preallocation Future is starving for lock on EntryLoggerAllocator.this to execute allocateNewLog", can you show a test case around that?